### PR TITLE
Akka http fix

### PR DESCRIPTION
**Pull request by tspring**
Transaction context can get lost between Directives because of Akka-Https use CallbackRunnables. This change attempts to propagate that context by weaving the CallbackRunnable constructor to create a token and link it on the run() method.

Due to performance concerns, I added a trackCallbackRunnable api to the Tracer, so that we only create/link the token when coming directly from AkkaHttp. We may consider putting this behind a config and/or traversing the tracer stack and not just checking the current tracer.

Thoughts?

### CODE REVIEW

File: instrumentation/scala-2.12.0/src/main/java/scala/concurrent/impl/CallbackRunnable_Instrumentation.java [hunk 1]
```diff
@@ -25,15 +29,31 @@
...
     /**
      * Override the setter for "value" so we can replace it with our wrapped version
      */
+
+    @Trace(async = true, excludeFromTransactionTrace = true)
     public void value_$eq(Try<T> original) {
         Weaver.callOriginal();
-
+        if (token != null) {
+            token.linkAndExpire();
```
Comment (breedx-nr) [inline instrumentation/scala-2.12.0/src/main/java/scala/concurrent/impl/CallbackRunnable_Instrumentation.java:54]: Is it safe for multiple threads to link and expire at the same time?  If not, there might be a race condition here and we may need to introduce a semaphore or other concurrent safety net.

File: newrelic-agent/src/main/java/com/newrelic/agent/tracers/AbstractTracer.java [hunk 2]
```diff
@@ -50,6 +50,7 @@
     private String customPrefix = "Custom";
     // doesn't need to be thread safe since this flag affects the decision to registerAsync
     private Boolean trackChildThreads = null;
+    private Boolean trackCallBackRunnable = false;
```
Comment (breedx-nr) [inline newrelic-agent/src/main/java/com/newrelic/agent/tracers/AbstractTracer.java:53]: `AtomicBoolean` here to help with threading issues?

File: newrelic-agent/src/main/java/com/newrelic/agent/tracers/AbstractTracer.java [hunk 3]
```diff
@@ -231,6 +232,24 @@ public boolean trackChildThreads() {
...
+        this.trackCallBackRunnable = shouldTrack;
+    }
+
+    @Override
+    public boolean trackCallbackRunnable() {
+        if (!this.trackCallBackRunnable) {
+            TracedMethod parent = this.getParentTracedMethod();
+            if (null == parent) {
+                return false;
+            } else {
```
Comment (breedx-nr) [inline newrelic-agent/src/main/java/com/newrelic/agent/tracers/AbstractTracer.java:246]: remove redundant `else`

File: newrelic-agent/src/main/java/com/newrelic/agent/tracers/AbstractTracer.java [hunk 4]
```diff
@@ -231,6 +232,24 @@ public boolean trackChildThreads() {
...
     }
 
+    @Override
+    public void setTrackCallbackRunnable(boolean shouldTrack) {
+        this.trackCallBackRunnable = shouldTrack;
+    }
+
+    @Override
+    public boolean trackCallbackRunnable() {
+        if (!this.trackCallBackRunnable) {
```
Comment (breedx-nr) [inline newrelic-agent/src/main/java/com/newrelic/agent/tracers/AbstractTracer.java:242]: If you factor out the guts of the `if` into a new private method called 'parentTrackCallbackRunnable()` or similar, then this collapses to a one-liner like
```
    @Override
    public boolean trackCallbackRunnable() {
        return this.trackCallBackRunnable || this.parentTrackCallbackRunnable();
    }
```

**Review by breedx-nr — CHANGES_REQUESTED** [review]
Looks pretty good.  We talked before about the idea that callbackrunnable sneaking all the way into the tracer feels a bit heavy handed, but it seems workable.  I also like that there is a config check, but did I miss where that config was added?

File: instrumentation/scala-2.12.0/src/main/java/scala/concurrent/impl/CallbackRunnable_Instrumentation.java (hunk 1 above)
Comment (tspring) [reply instrumentation/scala-2.12.0/src/main/java/scala/concurrent/impl/CallbackRunnable_Instrumentation.java:54]: Callbackrunnables can only be executed once, so this shouldn't be an issue

File: instrumentation/scala-2.12.0/src/main/java/scala/concurrent/impl/CallbackRunnable_Instrumentation.java [hunk 5]
```diff
@@ -25,15 +29,31 @@
 @Weave(originalName = "scala.concurrent.impl.CallbackRunnable")
 public class CallbackRunnable_Instrumentation<T> {
     private final Function1<Try<?>, Object> onComplete = Weaver.callOriginal();
-
     private Try<T> value = Weaver.callOriginal();
+    @NewField
+    private Token token;
```
[the remark this answers is not in the document]
Comment (tspring) [reply instrumentation/scala-2.12.0/src/main/java/scala/concurrent/impl/CallbackRunnable_Instrumentation.java:34]: NewFields are extension classes, not actual Java fields so this should be ok

File: newrelic-agent/src/main/java/com/newrelic/agent/tracers/AbstractTracer.java (hunk 2 above)
Comment (tspring) [reply newrelic-agent/src/main/java/com/newrelic/agent/tracers/AbstractTracer.java:53]: I don't think this needs to be thread safe as the tracer/TransactionActivity are thread specific

**Review by breedx-nr — APPROVED** [review]
Looks good. Thanks for the explanation and changes.

**Pull request merged**